### PR TITLE
Allocate HIR on an arena 1/4

### PR DESCRIPTION
**Pull request by cjgillot**
This PR is the first in a series of 4, aiming at allocating the HIR on an arena, as a memory optimisation.

1. This first PR lays the groundwork and migrates some low-hanging fruits.
2. The second PR will migrate `hir::Expr`, `hir::Pat` and related.
3. The third PR will migrate `hir::Ty` and related.
4. The final PR will be dedicated to eventual cleanups.

In order to make the transition as gradual as possible, some lowering routines receive `Box`-allocated data and move it into the arena. This is a bit wasteful, but hopefully temporary.
Nonetheless, special care should be taken to avoid double arena allocations.

Work mentored by @Zoxc.

### CODE REVIEW

**Comment by matthewjasper** [comment]
cc @rust-lang/compiler

File: src/librustc/arena.rs [hunk 1]
```diff
@@ -123,6 +122,10 @@ macro_rules! arena_types {
             [few] crate_variances: rustc::ty::CrateVariancesMap<'tcx>,
             [few] inferred_outlives_crate: rustc::ty::CratePredicatesMap<'tcx>,
             [] upvars: rustc_data_structures::fx::FxIndexMap<rustc::hir::HirId, rustc::hir::Upvar>,
+            // HIR nodes arenas
```
Comment (Zoxc) [inline src/librustc/arena.rs]: Maybe call this `HIR types` and put a newline above it to visually separate it from the rest.

File: src/librustc/hir/intravisit.rs [hunk 2]
```diff
@@ -382,10 +382,10 @@ pub trait Visitor<'v>: Sized {
 }
 
 /// Walks the contents of a crate. See also `Crate::visit_all_items`.
-pub fn walk_crate<'v, V: Visitor<'v>>(visitor: &mut V, krate: &'v Crate) {
+pub fn walk_crate<'v, V: Visitor<'v>>(visitor: &mut V, krate: &'v Crate<'v>) {
     visitor.visit_mod(&krate.module, krate.span, CRATE_HIR_ID);
-    walk_list!(visitor, visit_attribute, &krate.attrs);
-    walk_list!(visitor, visit_macro_def, &krate.exported_macros);
+    walk_list!(visitor, visit_attribute, krate.attrs);
```
Comment (Zoxc) [inline src/librustc/hir/intravisit.rs]: I'm curious, why didn't `&krate.attrs` still work?

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -383,19 +389,19 @@ impl<'a, 'b> Visitor<'a> for ImplTraitTypeIdVisitor<'b> {
...
+impl<'a, 'hir> LoweringContext<'a, 'hir> {
+    fn lower_crate(mut self, c: &Crate) -> hir::Crate<'hir> {
         /// Full-crate AST visitor that inserts into a fresh
         /// `LoweringContext` any information that may be
         /// needed from arbitrary locations in the crate,
         /// e.g., the number of lifetime generic parameters
         /// declared for every type and trait definition.
-        struct MiscCollector<'tcx, 'interner> {
-            lctx: &'tcx mut LoweringContext<'interner>,
+        struct MiscCollector<'tcx, 'interner, 'hir> {
```
Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: The names of these lifetimes are quite misleading. I think they should be named `MiscCollector<'misc, 'lowering, 'hir>` instead.

File: src/librustc/hir/lowering.rs [hunk 4]
```diff
@@ -2758,9 +2765,15 @@ impl<'a> LoweringContext<'a> {
     fn lower_pat(&mut self, p: &Pat) -> P<hir::Pat> {
         let node = match p.kind {
             PatKind::Wild => hir::PatKind::Wild,
-            PatKind::Ident(ref binding_mode, ident, ref sub) => {
-                let lower_sub = |this: &mut Self| sub.as_ref().map(|x| this.lower_pat(x));
-                self.lower_pat_ident(p, binding_mode, ident, lower_sub)
```
Comment (Centril) [inline src/librustc/hir/lowering.rs]: Why this change?

File: src/librustc/hir/lowering/item.rs [hunk 5]
```diff
@@ -25,11 +25,11 @@ use syntax_pos::Span;
 
 use rustc_error_codes::*;
 
-pub(super) struct ItemLowerer<'tcx, 'interner> {
-    pub(super) lctx: &'tcx mut LoweringContext<'interner>,
+pub(super) struct ItemLowerer<'tcx, 'interner, 'hir> {
```
Comment (Zoxc) [inline src/librustc/hir/lowering/item.rs]: This could be named `ItemLowerer<'a, 'lowering, 'hir>`.

File: src/librustc/hir/intravisit.rs (hunk 2 above)
Comment (cjgillot) [reply src/librustc/hir/intravisit.rs]: I changed the `HirVec<T>` to be `&'hir [T]`. Using `&krate.attrs` used to make a `&[T]`, and now makes a `&&[T]`.

File: src/librustc/hir/print.rs [hunk 6]
```diff
@@ -107,7 +107,7 @@ pub const INDENT_UNIT: usize = 4;
 /// it can scan the input text for comments to copy forward.
 pub fn print_crate<'a>(cm: &'a SourceMap,
                        sess: &ParseSess,
-                       krate: &hir::Crate,
+                       krate: &hir::Crate<'a>,
```
Comment (Zoxc) [inline src/librustc/hir/print.rs]: This could probably be `&hir::Crate<'_>`?

File: src/librustc/hir/lowering/item.rs [hunk 7]
```diff
@@ -282,29 +284,31 @@ impl LoweringContext<'_> {
...
                 hir::ItemKind::Const(
-                    self.lower_ty(
-                        t,
-                        if self.sess.features_untracked().impl_trait_in_bindings {
-                            ImplTraitContext::OpaqueTy(None)
-                        } else {
-                            ImplTraitContext::Disallowed(ImplTraitPosition::Binding)
-                        }
-                    ),
+                    self.arena.alloc(ty.into_inner()),
```
Comment (Centril) [inline src/librustc/hir/lowering/item.rs]: This feels noisy and repetitive. Could we bake `self.arena.alloc(ty.into_inner())` into `lower_ty` or if not, for whatever reason, add a `lower_ty_arena`?

Comment (Zoxc) [reply src/librustc/hir/lowering/item.rs]: I'd expect that to happen on the 3rd PR which moves `Ty` to the arena. I guess @cjgillot moved the fields of each type to the arena too so these show up as temporary changes.

File: src/librustc/arena.rs [hunk 8]
```diff
@@ -125,7 +125,12 @@ macro_rules! arena_types {
             // HIR nodes arenas
             [few] hir_forest: rustc::hir::map::Forest<$tcx>,
             [] attribute: syntax::ast::Attribute,
+            [] global_asm: rustc::hir::GlobalAsm,
```
Comment (Zoxc) [inline src/librustc/arena.rs]: This could have a `few` attribute since they're likely to be rare.

File: src/librustc/arena.rs [hunk 9]
```diff
@@ -123,6 +122,10 @@ macro_rules! arena_types {
             [few] crate_variances: rustc::ty::CrateVariancesMap<'tcx>,
             [few] inferred_outlives_crate: rustc::ty::CratePredicatesMap<'tcx>,
             [] upvars: rustc_data_structures::fx::FxIndexMap<rustc::hir::HirId, rustc::hir::Upvar>,
+            // HIR nodes arenas
+            [few] hir_forest: rustc::hir::map::Forest<$tcx>,
+            [] attribute: syntax::ast::Attribute,
+            [] macro_def: rustc::hir::MacroDef,
```
Comment (Zoxc) [inline src/librustc/arena.rs]: This should probably have a `few` attribute too.

File: src/librustc/hir/lowering/item.rs [hunk 10]
```diff
@@ -701,12 +701,12 @@ impl LoweringContext<'_, 'hir> {
...
     }
 
-    fn lower_foreign_item(&mut self, i: &ForeignItem) -> hir::ForeignItem {
+    fn lower_foreign_item(&mut self, i: &ForeignItem) -> hir::ForeignItem<'hir> {
         let def_id = self.resolver.definitions().local_def_id(i.id);
         hir::ForeignItem {
             hir_id: self.lower_node_id(i.id),
             ident: i.ident,
-            attrs: self.lower_attrs(&i.attrs),
+            attrs: self.lower_attrs_extendable(&i.attrs),
```
Comment (Zoxc) [inline src/librustc/hir/lowering/item.rs]: This seems a bit odd. Shouldn't `lower_attrs_extendable` still return a `Vec`?

File: src/librustc/hir/lowering.rs [hunk 11]
```diff
@@ -990,15 +990,14 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
         }
     }
 
-    fn lower_attrs_extendable(&mut self, attrs: &[Attribute]) -> Vec<Attribute> {
-        attrs
-            .iter()
-            .map(|a| self.lower_attr(a))
-            .collect()
+    fn lower_attrs_extendable(&mut self, attrs: &[Attribute]) -> &'hir [Attribute] {
```
Comment (Zoxc) [inline src/librustc/hir/lowering.rs]: This should still return `Vec<Attribute>` since callers may want to extend it.

Comment (Zoxc) [reply src/librustc/hir/lowering.rs]: `lower_attrs` is the function which should return `&'hir [Attribute]`.

File: src/librustc/hir/lowering/item.rs [hunk 12]
```diff
@@ -1042,7 +1042,7 @@ impl LoweringContext<'_, 'hir> {
 
     fn lower_body(
         &mut self,
-        f: impl FnOnce(&mut LoweringContext<'_, '_>) -> (HirVec<hir::Param>, hir::Expr),
+        f: impl for<'tcx> FnOnce(&mut LoweringContext<'_, 'tcx>) -> (&'tcx [hir::Param], hir::Expr),
```
Comment (Zoxc) [inline src/librustc/hir/lowering/item.rs]: We don't want to use `for<'tcx>` here.

File: src/librustc/hir/lowering.rs (hunk 11 above)
Comment (Zoxc) [reply src/librustc/hir/lowering.rs]: I suggest you add a `lower_attrs_arena` variant which returns `&'hir [Attribute]` which can be removed when all the attributes are migrated over.

File: src/librustc/hir/lowering/item.rs (hunk 7 above)
Comment (Zoxc) [reply src/librustc/hir/lowering/item.rs]: There doesn't seem to be that many fields that's moved to `&'hir Ty` though. I'd just keep `P<Ty>` for this PR to avoid unnecessary changes.

File: src/librustc/hir/lowering.rs (hunk 4 above)
Comment (cjgillot) [reply src/librustc/hir/lowering.rs]: This was a borrow checking workaround. It has been undone.

File: src/librustc/hir/lowering/item.rs (hunk 7 above)
Comment (cjgillot) [reply src/librustc/hir/lowering/item.rs]: Indeed, a lot of those disappear in #66942. I plan to do the final cleanup in PR 4/4.

**Comment by cjgillot** [comment]
Rebased with review comments.

**Review by eddyb — APPROVED** [review]
r=me if no concerns remain from @rust-lang/compiler (also, feel free to ping/PM me on Discord or Zulip in the future, when each PR is ready)

**Comment by nikomatsakis** [comment]
My take is that we should land this. Maybe we can nominate for a quick 👍 at triage meeting? Seems like it would be good to get some kind of perf numbers, too. Let's kick that off.

@bors try

**Comment by nikomatsakis** [comment]
@bors try- 

Never mind, @eddyb points out it'd be better to test the final PR https://github.com/rust-lang/rust/pull/67032

**Comment by alex** [comment]
FYI, per https://buildbot2.rust-lang.org/homu/queue/all bors seems to think this PR is being tested, but I don't think it is?

**Comment by Zoxc** [comment]
I do expect the performance impact of these PRs to be very similar to https://github.com/rust-lang/rust/pull/57173 (I think it was 1-3% improvements). It's possible that there's types with destructors left in these PRs though since they don't make the HIR types implement `Copy` which could impact performance.

File: src/librustc/hir/lowering/item.rs (hunk 7 above)
Comment (cjgillot) [reply src/librustc/hir/lowering/item.rs]: After discussing with @pnkfelix, I remembered another reason to migrate eagerly the `P<Ty>` into `&'hir Ty`. If I do not do that, I get unused lifetime parameter errors in intermediate commits (ex: `TraitItemKind`). This keeps me from splitting in compiling commits.

**Comment by michaelwoerister** [comment]
I looked through the changes here a little bit and this looks good to me in general. Allocating the HIR in arenas is a good idea. Thanks for putting in all this effort, @cjgillot!

Performance looks good too. I'm wondering though if we could do even better. As it is now (iirc) every type has it's own arena, which means that pieces of HIR that are close to each other will probably end up being wide apart in memory. Trying to allocate different types from the same arena in HIR-lowering order would probably improve this. I'm not sure if there are technical difficulties with that (the single-threaded, non-destructors case sounds straightforward). Maybe this is something to look into in the future.

**Comment by Zoxc** [comment]
@michaelwoerister All the types without destructors will end up in the same arena which should be most of the HIR at the end of this PR set.

**Comment by Centril** [comment]
I want to highlight that a really great effect of this PR is that we can cleanup a bunch of code due to better pattern matching since `P<T>` isn't interfering anymore. Maybe we should file an issue for the cleanup?

**Comment by michaelwoerister** [comment]
@Zoxc That sounds like we want then. Thanks for pointing it out.

**Comment by cjgillot** [comment]
@Centril a lot of this cleanup is done in the 4th PR, where `P`is removed altogether.

In the mean time, is there still anything blocking approval? I got a bit lost in all the comments.

**Comment by Centril** [comment]
@cjgillot Oh, that is just lovely, thanks! ❤️ 

👍 for merging on my end, but I'm also not the technical reviewer :)

**Comment by matthewjasper** [comment]
Blocked on beta branching

**Comment by cjgillot** [comment]
Rebased.
